kpet/cmd_arch.py

```diff
@@ -41,9 +46,18 @@ def main(args):
     database = data.Base(args.db)
     if args.action == 'list':
         regex = re.compile(args.regex or ".*")
-        for arch in sorted(database.arches):
-            # filter using regex; if it's not set, print all
-            if regex.fullmatch(arch):
-                print(arch)
+        tree_regex = re.compile(args.tree or ".*")
+        arch_names = set()
+        filtered = list(value['arches'] for (key, value)
+                        in database.trees.items()
+                        if tree_regex.fullmatch(key))
+        # Flattens the list of lists and removes duplicates
+        filtered = {y for x in filtered for y in x}
+
+        for arch in list(filter(regex.fullmatch, filtered)):
+            arch_names.add(arch)
+
+        for name in sorted(arch_names):
+            print(name)
```

Comment:
These two for loops would probably be better to just replace with
```
for arch in sorted(list(filter(regex.fullmatch, filtered))):
    print(arch)
```
And the "arch_names" variable could be removed entirely